scripts/ssl_tts/ssl_tts_vc.py

```diff
@@ -0,0 +1,254 @@
+    # print("signal batch", signal_batch.shape)
+    signal_length_batch = torch.stack([torch.tensor(signal_batch.shape[1]) for _i in range(len(all_segments))])
+    # print("signal length", signal_length_batch.shape)
+    signal_batch = signal_batch.to(device)
+    signal_length_batch = signal_length_batch.to(device)
+    _, speaker_embeddings, _, _, _ = ssl_model.forward_for_export(
+        input_signal=signal_batch, input_signal_length=signal_length_batch, normalize_content=True
+    )
+
+    speaker_embedding = torch.mean(speaker_embeddings, dim=0)
+    l2_norm = torch.norm(speaker_embedding, p=2)
+    speaker_embedding = speaker_embedding / l2_norm
+
+    return speaker_embedding[None]
+
+
+def get_ssl_features_disentsngled(
+    ssl_model, wav_featurizer, audio_path, emb_type="embedding_and_probs", use_unique_tokens=False, device="cpu"
+):
+    wav = load_wav(audio_path, wav_featurizer)
+    audio_signal = wav[None]
+    audio_signal_length = torch.tensor([wav.shape[0]])
+    audio_signal = audio_signal.to(device)
+    audio_signal_length = audio_signal_length.to(device)
+    _, speaker_embedding, content_embedding, content_log_probs, encoded_len = ssl_model.forward_for_export(
+        input_signal=audio_signal, input_signal_length=audio_signal_length, normalize_content=True
+    )
+
+    content_embedding = content_embedding[0, : encoded_len[0].item()]
+    content_log_probs = content_log_probs[: encoded_len[0].item(), 0, :]
+    content_embedding = content_embedding.t()
+    content_log_probs = content_log_probs.t()
+    content_probs = torch.exp(content_log_probs)
+
+    if emb_type == "probs":
+        final_content_embedding = content_probs
+
+    elif emb_type == "embedding":
+        final_content_embedding = content_embedding
+
+    elif emb_type == "log_probs":
+        final_content_embedding = content_log_probs
+
+    elif emb_type == "embedding_and_probs":
+        final_content_embedding = torch.cat([content_embedding, content_probs], dim=0)
+
+    duration = torch.ones(final_content_embedding.shape[1]) * 4.0
+    if use_unique_tokens:
+        token_predictions = torch.argmax(content_probs, dim=0)
+        # print("token predictions:", token_predictions)
+        content_buffer = [final_content_embedding[:, 0]]
+        unique_content_embeddings = []
+        unique_tokens = []
+        durations = []
+        for _t in range(1, final_content_embedding.shape[1]):
+            if token_predictions[_t] == token_predictions[_t - 1]:
+                content_buffer.append(final_content_embedding[:, _t])
+            else:
+                durations.append(len(content_buffer) * 4)
+                unique_content_embeddings.append(torch.mean(torch.stack(content_buffer), dim=0))
+                content_buffer = [final_content_embedding[:, _t]]
+                unique_tokens.append(token_predictions[_t].item())
+
+        if len(content_buffer) > 0:
+            durations.append(len(content_buffer) * 4)
```

Comment:
Can these `4`s be replaced with a user-provided param (or one saved in the model)?

Reply:
updated